Fix error while accessing tables when using Glue

pratyakshsharma opened the pull request:
OpenCSVSerde deserializes columns from csv file into strings only. Currently the CSV support in Presto relies on metastore to return all data types as string. Glue metastore returns the original data type which results in error when trying to cast to string.

Fixes https://github.com/prestodb/presto/issues/15869.

Test plan - (Please fill in how you tested your changes)

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines. Don't forget to follow our [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution) for any code copied from other projects.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== NO RELEASE NOTE ==
```

imjalpreet reviewed (changes requested):
Thank you for the fix. Most of the comments have been addressed offline. I did another pass and just have a minor comment. I will let you know if I have any other.

Another request: can we change the commit message to reflect error while accessing CSV tables.

pratyakshsharma commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1755,7 +1755,7 @@ public static boolean shouldCreateFilesForMissingBuckets(Table table, ConnectorS
     }
 
-    private MetastoreContext getMetastoreContext(ConnectorSession session)
+    MetastoreContext getMetastoreContext(ConnectorSession session)
```
got you.

yzhang1991 commented:
@imjalpreet can you follow up on this? I heard you have a review comment that you need to add here. Thanks!

imjalpreet commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/ExtendedHiveMetastore.java:
```diff
@@ -77,7 +77,7 @@
-    Optional<Partition> getPartition(MetastoreContext metastoreContext, String databaseName, String tableName, List<String> partitionValues);
+    Optional<Partition> getPartition(MetastoreContext metastoreContext, Table table, List<String> partitionValues);
 
     Optional<List<String>> getPartitionNames(MetastoreContext metastoreContext, String databaseName, String tableName);
```
Since we are introducing table as an argument to other relevant methods, I think we can also make a similar change for `getPartitionNames`. This will help in reducing multiple `getTable` calls when called from `SemiTransactionalHiveMetastore` since we already fetch the table object there.

pratyakshsharma replied:
This is a valid ask. However, MockHiveMetastoreClient gets called from TestCachingHiveMetastore tests internally and this client is implemented in a way that it checks for BAD_DATABASE for all get calls. For example, `testInvalidGetPartitionNames` in TestCachingHiveMetastore will ultimately call `getPartitionNames` in MockHiveMetastoreClient which verifies database name. 

Having said that, I have a plan in my mind. To incorporate the optimisation that you proposed, we need to remove db name check from getTable() method in MockHiveMetastoreClient (https://github.com/prestodb/presto/blob/1c9de99de7cb65b97f0bb304c3e58fb4caf02dac/presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/thrift/MockHiveMetastoreClient.java#L165) and remove the corresponding test case from TestCachingHiveMetastore (https://github.com/prestodb/presto/blob/1c9de99de7cb65b97f0bb304c3e58fb4caf02dac/presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestCachingHiveMetastore.java#L153). This essentially means MockHiveMetastoreClient will always return a table object irrespective of the database name passed. But this is only a test case and should not cause any issues as such. Once this is done, we can do the changes that you proposed @imjalpreet .

Good to get some committers eyes on this as well. @tdcmeehan @prestodb/committers @NikhilCollooru @highker .

pratyakshsharma replied:
@prestodb/committers I pushed the second commit separately to implement what I proposed in the above comment. Feel free to have a look and share your thoughts on this.

imjalpreet reviewed (commented):
@pratyakshsharma Let me know what you think about this.

yzhang1991 commented:
@imjalpreet can you check this and give your explicit approval if everything looks good? Thanks!

pratyakshsharma commented:
@imjalpreet gentle ping!

imjalpreet reviewed (changes requested):
@pratyakshsharma thanks for the changes, I have done an initial review, and will add more if something else comes to my mind.

Also, I think we can rebase with master since it shows there are some conflicts.

NikhilCollooru commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -797,10 +799,15 @@ public Map<String, Optional<Partition>> getPartitionsByNames(MetastoreContext me
     {
+        String databaseName = partitionName.getKey().getHiveTableName().getDatabaseName();
+        String tableName = partitionName.getKey().getHiveTableName().getTableName();
+        Optional<Table> table = getTable(partitionName.getContext(), databaseName, tableName);
```
this will result an additional getTable roundtrip call to metastore ? this might cause performance regression

pratyakshsharma replied:
this should get it from the cache in case it is fetched already. So I guess there should not be any performance regression. @NikhilCollooru

NikhilCollooru replied:
But Table cache is not enabled by default right. So if we dont have cache enabled it would result in an additional call ?

pratyakshsharma replied:
yes, table cache is not enabled by default. But we actually need table object in GlueToPrestoConverter constructor call to be able to get serde object and fix the original issue reported in https://github.com/prestodb/presto/issues/15869. I cannot think of any way to get the serde without getting the table object. 
I am open hear your suggestions around the same. @NikhilCollooru

NikhilCollooru replied:
I am wondering if you can gate the code changes in this PR behind a config. So that we enable this code path only when one needs to support accessing tables using Glue.

pratyakshsharma replied:
The code changes in this PR involve API signature changes in ExtendedHiveMetastore and are present across multiple files. Also the cache structure got modified in CachingHiveMetastore. Hence guarding these changes behind a config seems a tedious task and will make the code too dirty. 
Also I am not sure how many users are actually using Presto without enabling cache. Most users will tend to use Presto with Caching enabled I believe. One workaround I can think of is enabling cache by default with some sane values for expiry and max size. I guess this should not be a breaking change for users, since Presto already has the concept of open and closed partitions. 
Again I am open to hearing your thoughts on this @NikhilCollooru

NikhilCollooru replied:
After reading code i see that we memoize/cache all calls to Metastore  within a transaction. Meaning getTable on a particular table will go to Metastore only once and duplicate getTable future calls on that same table will be served from the cache. This is irrespective of whether table cache is enabled or not.
https://github.com/prestodb/presto/blob/master/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L160

@pratyakshsharma can you run some tests and confirm that the actual number of getTable calls to Metastore for a given query (not involving glue tables) will remain same before and after this PR ?

pratyakshsharma replied:
Sure, will get back to you on this.

pratyakshsharma replied:
As I checked, it looks like SemiTransactionalHiveMetastore has issues with creating CachingHiveMetastore instance properly. With metastore types other than `alluxio`, CachingHiveMetastore has another CachingHiveMetastore as delegate, which results in routing the calls again to CachingHiveMetastore. Due to this, 2 extra getTable calls were being made to metastore. Here is the plan to fix this - 
1. Fix creation of CachingHiveMetastore instance. (this removes 1 extra call)
2. Reintroduce api getPartitionsByNames which takes databaseName and tableName as parameters, and call it in case of thrift hive metastore. (this removes 1 extra call). 

I will be taking care of this fix in my next commit @NikhilCollooru

pratyakshsharma replied:
@NikhilCollooru please take a pass. I can confirm the number of getTable calls to metastore do not increase after my changes. In fact, one redundant getTable call to CachingHiveMetastore is now reduced.

pratyakshsharma commented on presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/thrift/MockHiveMetastoreClient.java:
```diff
@@ -66,13 +66,16 @@
     public static final String TEST_TABLE_WITH_CONSTRAINTS = "testtbl_constraints";
     public static final Map<String, List<FieldSchema>> SCHEMA_MAP = ImmutableMap.of(
             TEST_DATABASE + TEST_TABLE, ImmutableList.of(new FieldSchema("key", "string", null)),
+            BAD_DATABASE + TEST_TABLE, ImmutableList.of(new FieldSchema("key", "string", null)),
```
This was actually needed for fixing some test case. If you want me to provide more details, I can check again and share the test case for which this was needed.

pratyakshsharma replied:
Basically if this entry is not added here, NPE is thrown at linehttps://github.com/prestodb/presto/blob/791ff32cd325a0e11e6868f69afac81e1bb271ee/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java#L438. 

I will have to add a check at this place and that would make the code more dirty. So I guess I am unable to think of anything right now for your comment.

imjalpreet replied:
We should be able to remove this entry with the new approach. Let me know if we face any issues.

pratyakshsharma commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestCachingHiveMetastore.java:
```diff
@@ -150,8 +158,6 @@ public void testGetTable()
 
     public void testInvalidDbGetTable()
     {
```
you are right, removing this test for now.

pratyakshsharma commented:
@imjalpreet @NikhilCollooru this should be good for another pass.

pratyakshsharma commented:
@imjalpreet ping!

imjalpreet commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestCachingHiveMetastore.java:
```diff
@@ -148,35 +156,28 @@ public void testGetTable()
         assertEquals(mockClient.getAccessCount(), 2);
     }
 
-    public void testInvalidDbGetTable()
```
Instead of removing it, can we try to still test this with the updated changes? Do you see an issue while adding this test?

pratyakshsharma replied:
Added this test back using another table variable - `RANDOM_TABLE`. This was the easiest way out without breaking anything and for keeping this test case in place. Let me know if you think otherwise.

imjalpreet reviewed (changes requested):
I have added some minor comments, let me know in case you have any questions.

pratyakshsharma commented:
@imjalpreet Please take a pass when you have cycles.

imjalpreet commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -675,14 +679,18 @@ public Optional<Partition> getPartition(MetastoreContext metastoreContext, Strin
     {
-        return delegate.getPartitionNames(hiveTableNameKey.getContext(), hiveTableNameKey.getKey().getDatabaseName(), hiveTableNameKey.getKey().getTableName());
+        Optional<Table> table = getTable(hiveTableNameKey.getContext(), hiveTableNameKey.getKey().getDatabaseName(), hiveTableNameKey.getKey().getTableName());
+        if (!table.isPresent()) {
```
We should try to move this check to the respective metastore implementation since Thrift and Glue both implementations behave differently in the current scenario. ThriftHiveMetastore doesn't throw an exception whereas GlueHiveMetastore does.

pratyakshsharma replied:
This requires making change in getPartitionNames method to pass Optional<Table> instead of Table as an argument. But this is a valid ask and hence I will change the method signature.

imjalpreet commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestCachingHiveMetastore.java:
```diff
@@ -162,21 +172,23 @@ public void testGetPartitionNames()
     public void testInvalidGetPartitionNames()
     {
-        assertEquals(metastore.getPartitionNames(TEST_METASTORE_CONTEXT, BAD_DATABASE, TEST_TABLE).get(), ImmutableList.of());
+        assertEquals(metastore.getPartitionNames(TEST_METASTORE_CONTEXT, getTableOrElseThrow(BAD_DATABASE, TEST_TABLE)).get(), ImmutableList.of());
```
We can call getTable since we don't want to throw an error in this case if table doesn't exist.

pratyakshsharma replied:
As per the above comment, getPartitionNames anyways has to be changed, so this comment is taken care of.

imjalpreet commented on presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/thrift/MockHiveMetastoreClient.java:
```diff
@@ -162,13 +168,13 @@ public Table getTable(String dbName, String tableName)
         if (throwException) {
             throw new RuntimeException();
         }
-        if (!dbName.equals(TEST_DATABASE) || (!tableName.equals(TEST_TABLE) && !tableName.equals(TEST_TABLE_WITH_CONSTRAINTS))) {
```
We can add this check back once we make the other suggested changes as discussed.

pratyakshsharma replied:
This cannot be added back because this will result in returning Optional.empty() from getTable() call and the testcase `testInvalidGetPartitionNames` calls getPartitionNames of CachingHiveMetastore, where HiveTableName object needs to be created as the caching key. Hence we need to get databaseName and tableName from `Optional<Table>` at this point.

imjalpreet replied:
As discussed offline, we can make changes to the CachingKey for partitionNamesCache to make use of the table object rather than HiveTableNameKey since we have the table object now as part of the method arguments.

imjalpreet commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestCachingHiveMetastore.java:
```diff
@@ -148,9 +157,10 @@ public void testGetTable()
     public void testInvalidDbGetTable()
     {
-        assertFalse(metastore.getTable(TEST_METASTORE_CONTEXT, BAD_DATABASE, TEST_TABLE).isPresent());
+        assertFalse(metastore.getTable(TEST_METASTORE_CONTEXT, BAD_DATABASE, RANDOM_TABLE).isPresent());
```
I think we should be able to change it back to TEST_TABLE with the new approach

pratyakshsharma replied:
Because of the above stated reason, TEST_DATABASE name check cannot be added in getTable method in MockHiveMetastoreClient. So this should continue to make use of RANDOM_TABLE constant.

imjalpreet reviewed (changes requested):
@pratyakshsharma Sorry for the delay, some minor review comments(hopefully last iteration). As discussed offline, let me know in case we face any issues with the suggested changes.

pratyakshsharma commented:
@imjalpreet Thank you for your valuable inputs which helped me in getting this PR to a better shape and also understand the presto codebase a little better. I have addressed all your review comments now, please have a look.

imjalpreet commented on presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/thrift/MockHiveMetastoreClient.java:
```diff
@@ -63,6 +63,9 @@
     public static final String TEST_TABLE = "testtbl";
+
+    // Table to use with BAD_DATABASE
+    public static final String RANDOM_TABLE = "randomtbl";
```
Unused variable, can be removed now.

imjalpreet commented on presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/thrift/MockHiveMetastoreClient.java:
```diff
@@ -167,8 +172,8 @@ public Table getTable(String dbName, String tableName)
-                TEST_TABLE,
-                TEST_DATABASE,
+                tableName,
+                dbName,
```
I think we don't need to change TEST_DATABASE to dbName, since we are not allowing any other dbName via the if condition above.

imjalpreet reviewed (changes requested):
Some minor comments, otherwise LGTM 👍🏼

pratyakshsharma commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestCachingHiveMetastore.java:
```diff
@@ -426,7 +444,7 @@ public void testListRoles()
     public void testInvalidGetPartitionsByNames()
     {
-        Map<String, Optional<Partition>> partitionsByNames = metastore.getPartitionsByNames(TEST_METASTORE_CONTEXT, BAD_DATABASE, TEST_TABLE, ImmutableList.of(TEST_PARTITION1));
+        Map<String, Optional<Partition>> partitionsByNames = metastore.getPartitionsByNames(TEST_METASTORE_CONTEXT, getTableOrElseThrow(TEST_DATABASE, TEST_TABLE), ImmutableList.of(BAD_PARTITION));
```
This change was made intentionally since this test case ideally should be checking for bad partition rather than checking with bad database. This is the same case in Trino as well.

imjalpreet reviewed (dismissed):
LGTM 👍 Thanks for the fix. Can we also squash the commits to have only the required commits.

pratyakshsharma commented:
@prestodb/committers Please take a pass, this should be good to merge.

pratyakshsharma commented:
@prestodb/committers ping!

yzhang1991 commented:
Hi @NikhilCollooru @rschlussel, can we take a look at this PR? Thanks!

NikhilCollooru commented:
reposting my earlier question from the comment. 

Looks like there will be an additional call to getTable inside loadPartitionsByName. But the table cache is not enabled by default. So it will be an increase in latency because of extra call. Is my understanding correct ?

pratyakshsharma commented:
> reposting my earlier question from the comment.

Looks like there will be an additional call to getTable inside loadPartitionsByName. But the table cache is not enabled by default. So it will be an increase in latency because of extra call. Is my understanding correct ?

Replied on your comment itself. @NikhilCollooru

pratyakshsharma commented:
@NikhilCollooru gentle ping! 

cc @prestodb/committers

pratyakshsharma commented:
@NikhilCollooru Pinging to see if you can review the latest changes anytime soon?

NikhilCollooru commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -159,18 +163,34 @@ public CachingHiveMetastore(
-                0.0,
-                partitionCacheMaxColumnCount,
-                NOOP_METASTORE_CACHE_STATS);
+        if (delegate instanceof AlluxioHiveMetastore) {
+            return new CachingHiveMetastore(
+                    delegate,
+                    newDirectExecutorService(),
+                    isMetastoreImpersonationEnabled,
+                    OptionalLong.empty(),
+                    OptionalLong.empty(),
+                    maximumSize,
+                    false,
+                    ALL,
+                    0.0,
+                    partitionCacheMaxColumnCount,
+                    NOOP_METASTORE_CACHE_STATS);
+        }
+        else {
+            return new CachingHiveMetastore(
+                    delegate.getDelegate().get(),
+                    newDirectExecutorService(),
+                    isMetastoreImpersonationEnabled,
+                    OptionalLong.empty(),
+                    OptionalLong.empty(),
+                    maximumSize,
+                    false,
+                    ALL,
+                    0.0,
+                    partitionCacheMaxColumnCount,
+                    NOOP_METASTORE_CACHE_STATS);
+        }
```
Can you explain a bit what you are doing here ?

pratyakshsharma replied:
In cases other than AlluxioHiveMetastore, the delegate getting passed to `memoizeMetastore()` method is already an instance of CachingHiveMetastore. This was resulting in CachingHiveMetastore calling CachingHiveMetastore again via its delegate, which in turn will then call the actual metastore implementation. 
Hence for cases other than AlluxioHiveMetastore, I am now doing `delegate.getDelegate().get()` so that the intended metastore is set as the delegate and there are no redundant calls to CachingHiveMetastore.
If you feel this is something not related to this PR, I can raise a new one for this and we can merge that first. I came to know about this issue when checking the number of getTable() calls. @NikhilCollooru

NikhilCollooru replied:
The purpose of memoize metastore is to return a CachingHiveMetastore with 100% caching enabled for all calls. 
CachingHiveMetastore(scope = ALL) --calls -> normal CachingHIveMetastore(scopre=as per what is configured in the config) ---calls --> actual Metastore implementation.

The reasoning is that the redundant calls to Metastore inside one transaction should be served from cache irrespective of the actual scope configured. Meaning lets assume scope=None is configured. This means no calls will be cached inside CachingHiveMetastore and all calls will be forwarded to actual Metastore. Lets say the query is using tableA 10 times in the query (select a from tableA union select b from tableA union ..etc) .. then the memoize metastore calls will make sure that only the first getTable(tableA) goes to actual Metastore and the remaining redundant 9 getTable calls will be served from the cache. 

Your current implementation is removing the middle CachingHiveMetastore and enabling 100% caching to all Metastore calls irrespective of scope configured. It is incorrect.

pratyakshsharma replied:
@NikhilCollooru I have pinged you on presto slack. Let us discuss this in detail there.

pratyakshsharma replied:
@NikhilCollooru I have pinged you on slack for deeper discussions. 
Just trying to understand your example here. 

> CachingHiveMetastore(scope = ALL) (this instance is per query/transaction)--calls -> normal CachingHIveMetastore(scopre=as per what is configured in the config)(this is singleton instance that is shared by all queries) ---calls --> actual Metastore implementation.

Lets call the per query/transaction CHM as CHM1 and the singleton instance as CHM2. CHM1 gets instantiated in memoizeMetastore call.
Lets assume scope=PARTITION is configured. As per the [code](https://github.com/prestodb/presto/blob/c3dbf9a6655a36321dc595222f256f5e52889547/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L166), CHM1 always has ttl = 0. So in your example, tableCache object for CHM1 has ttl=0 and since scope is PARTITION, tableCache object for CHM2 also has ttl=0. So every getTable call goes to actual metastore only. 

Please let me know if I am missing anything here.

pratyakshsharma replied:
Having said that, on a high level, I still feel 100% caching within a given transaction can be achieved with a single instance of CHM. All you need to do is only allow scope=ALL and enable CHM1 to make use of MetastoreClientConfig configs. 
Also if we ultimately want to have 100% caching enabled for all calls within a given transaction/query, then what is the point of providing the option of setting scope to PARTITION? Please let me know @NikhilCollooru

NikhilCollooru replied:
> Lets assume scope=PARTITION is configured. As per the [code](https://github.com/prestodb/presto/blob/c3dbf9a6655a36321dc595222f256f5e52889547/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L166), CHM1 always has ttl = 0.
This is incorrect. OptionalLong.empty() **DOES NOT** mean ttl=0 . Please see the usage here [https://github.com/prestodb/presto/blob/master/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L1093](code). It means TTL is not set. In guava cache, it means that the cache entry lives until the max capacity of the cache (1000 here) is reached.
> By default cache instances created by CacheBuilder will not perform any type of eviction.
https://guava.dev/releases/17.0/api/docs/com/google/common/cache/CacheBuilder.html

NikhilCollooru replied:
> Also if we ultimately want to have 100% caching enabled for all calls within a given transaction/query, then what is the point of providing the option of setting scope to PARTITION?

let me re-iterate cache scope is for sharing the cache across all queries. When we set the scope to PARTITION ..it means that the partition object is cache across all queries. If CHM2 does not exist ...then there is no way to share the cache partition/table object across queries .

pratyakshsharma replied:
Thank you for the useful explanations @NikhilCollooru and thank you for being patient with my doubts. I now fully understand the use of CHM2, but I still think the overall design can be simplified. I guess providing the option of setting scope=PARTITION is redundant, because the partition caches are already taken care of with scope=ALL.

I am trying to understand the use of CHM1 now, because the possible values of scope are `ALL` and `PARTITION` only. Setting any other value for scope (for example `NONE`) results in throwing IllegalArgumentException. Consider a scenario where only CHM2 exists with scope=ALL always. Effectively if I am removing this memoizeMetastore() method completely, do you see any case where 100% caching is not enabled for all calls in a transaction?

In other words, is CHM1 needed only because we are providing the option of setting scope=PARTITION to users, or does it solve any other use case as well? This should be my final question on this discussion. Looking forward to hearing from you.

NikhilCollooru replied:
> In other words, is CHM1 needed only because we are providing the option of setting scope=PARTITION to users, or does it solve any other use case as well?
Per-transaction metastore cache addresses the issue of repeated access of same data in the metastore. CHM1 will make sure that a given query will get consistent view of snapshot of metadata. It will make sure all the duplicate metastore calls will return the same data irrespective of whether the metadata changed on server side or not. Assume a given query calls metadataCall1 multiple times. CHM1 will make sure the result of the metadataCall1 is always the same across all the calls even if the data changed on the metastore side. 

lets say CHM1 does not exist and CHM2 has scope = NONE. Then every call goes to Metastore and its not guaranteed that the same call might return the same data..since there is no guarantee that the table/partition is not modified during the duration of the query.

pratyakshsharma replied:
So on a high level, CHM1 is needed to cater to cases which cannot be covered with CHM2. At present, setting scope=PARTITION seems to be the only such case to me because if the scope is set to ALL for CHM2, it can only cache every call. 
To re-iterate my point, scope cannot be set to NONE as per the [code](https://github.com/prestodb/presto/blob/864619df81857c61db353af99455369452771f8f/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L225), and I am still not convinced on why we provide the option of setting scope as PARTITION. 
As I understand this, our goal of caching every single metadata call in a transaction is only getting complicated if some user overrides the value of scope to PARTITION ([it is by default set to ALL](https://github.com/prestodb/presto/blob/864619df81857c61db353af99455369452771f8f/presto-hive-metastore/src/main/java/com/facebook/presto/hive/MetastoreClientConfig.java#L49)). As per the documentation, guava cache does not put any constraint on the maximum size as well, so having only CHM2 in place is again not a problem. 

Hence I suggest to change this caching design of presto to only have CHM2 in place. In case of Alluxio the singleton instance is not CachingHiveMetastore, but that is all implementation details. We can maybe take it up separately if you and the community agrees. I am reverting the changes for now so we can get this PR merged. @NikhilCollooru

pratyakshsharma replied:
Actually after thinking it through, I guess what you are saying makes sense. Both CHM1 and CHM2 are needed in place because if the scope of CHM2 is set to ALL, then all the metadata calls are cached including listTableNames, listSchemaNames in a shared cache, and this might not be the desired behaviour for many users. 

I am currently working on fixing the build and failing test cases. Once that is done, I will ping you again for another pass @NikhilCollooru

pratyakshsharma commented:
@NikhilCollooru Please take a pass. I have fixed the test cases and addressed all comments.

pratyakshsharma commented:
@NikhilCollooru gentle ping!

pratyakshsharma commented:
https://github.com/prestodb/presto/pull/20228

pratyakshsharma commented:
@NikhilCollooru please take a pass.

pratyakshsharma commented:
@NikhilCollooru gentle ping!

NikhilCollooru commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/ExtendedHiveMetastore.java:
```diff
@@ -89,7 +89,7 @@ default Optional<Table> getTable(MetastoreContext metastoreContext, HiveTableHan
     Optional<Partition> getPartition(MetastoreContext metastoreContext, Table table, List<String> partitionValues);
 
-    Optional<List<String>> getPartitionNames(MetastoreContext metastoreContext, String databaseName, String tableName);
+    Optional<List<String>> getPartitionNames(MetastoreContext metastoreContext, Optional<Table> table);
```
Why is it Optional ? Why do you want to call getPartitionNames if its empty ?

pratyakshsharma replied:
There is an existing check at both places in SemiTransactionalHiveMetastore to see if tableOptional is empty or not. So this method is only called when table object is not empty.

NikhilCollooru commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -648,7 +648,7 @@ private Optional<List<String>> doGetPartitionNames(
                 }
                 else {
-                    partitionNameResult = delegate.getPartitionNames(metastoreContext, hiveTableHandle.getSchemaName(), hiveTableHandle.getTableName());
+                    partitionNameResult = delegate.getPartitionNames(metastoreContext, table);
```
we know the table exists .. since isPresent check is already made in line 636. we can send table.get() as input.
Having an optional value for the main parameter of getPartitionNames doesnt make sense

pratyakshsharma replied:
The problem is getPartitionNames gets called from getPartitionNamesByFilter as well. If I change the method signature of getPartitionNames to remove Optional<Table> and have Table instead, that effectively shifts the empty table checks to other places, which is not desirable. Hence I would prefer keeping it as is.

NikhilCollooru reviewed (commented):
I have one main comment about having Optional<Table> as parameter. It doesnt make sense to have optional for that parameter. 
Also you can probably squash some of the commits .

pratyakshsharma commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:
```diff
@@ -107,6 +108,7 @@ public void testRowNumber()
         testRankingFunction("row_number()", true);
     }
 
+    @Ignore
```
sure.

pratyakshsharma commented:
@NikhilCollooru This should be good for final pass. Please take a look.

pratyakshsharma commented:
@NikhilCollooru ping.

pratyakshsharma commented:
@NikhilCollooru @tdcmeehan Can we close this soon?

pratyakshsharma commented:
@NikhilCollooru ping

tdcmeehan commented:
@NikhilCollooru I am going to take over this review.  Please, if you have any feedback, let us know.
